src/main/scala/UARTAdapter.scala

```diff
@@ -0,0 +1,149 @@
+
+  val uart_io = io.uart
+
+  val sTxIdle :: sTxWait :: sTxData :: sTxBreak :: Nil = Enum(4)
+  val txState = RegInit(sTxIdle)
+  val txData = Reg(UInt(dataWidth.W))
+  // iterate through bits in byte to deserialize
+  val (txDataIdx, txDataWrap) = Counter(txState === sTxData, dataWidth)
+  // iterate using div to convert clock rate to baud
+  val (txBaudCount, txBaudWrap) = Counter(txState === sTxWait, div)
+  val (txSlackCount, txSlackWrap) = Counter(txState === sTxIdle && uart_io.txd === 0.U, 4)
+
+  switch(txState) {
+    is(sTxIdle) {
+      when(txSlackWrap) {
+        txData  := 0.U
+        txState := sTxWait
+      }
+    }
+    is(sTxWait) {
+      when(txBaudWrap) {
+        txState := sTxData
+      }
+    }
+    is(sTxData) {
+      txData := txData | (uart_io.txd << txDataIdx)
+      when(txDataWrap) {
+        txState := Mux(uart_io.txd === 1.U, sTxIdle, sTxBreak)
+      }.otherwise {
+        txState := sTxWait
+      }
+    }
+    is(sTxBreak) {
+      when(uart_io.txd === 1.U) {
+        txState := sTxIdle
+      }
+    }
+  }
+
+  txfifo.io.enq.bits  := txData
+  txfifo.io.enq.valid := txDataWrap
+
+  val sRxIdle :: sRxStart :: sRxData :: Nil = Enum(3)
+  val rxState = RegInit(sRxIdle)
+  // iterate using div to convert clock rate to baud
+  val (rxBaudCount, rxBaudWrap) = Counter(true.B, div)
+  // iterate through bits in byte to deserialize
+  val (rxDataIdx, rxDataWrap) = Counter(rxState === sRxData && rxBaudWrap, dataWidth)
+
+  uart_io.rxd := 1.U
+  switch(rxState) {
+    is(sRxIdle) {
+      uart_io.rxd := 1.U
+      when (rxBaudWrap && rxfifo.io.deq.valid) {
+        rxState := sRxStart
+      }
+    }
+    is(sRxStart) {
+      uart_io.rxd := 0.U
+      when(rxBaudWrap) {
+        rxState := sRxData
+      }
+    }
+    is(sRxData) {
+      uart_io.rxd := (rxfifo.io.deq.bits >> rxDataIdx)(0)
+      when(rxDataWrap && rxBaudWrap) {
+        rxState := sRxIdle
+      }
+    }
+  }
+  rxfifo.io.deq.ready := (rxState === sRxData) && rxDataWrap && rxBaudWrap
+
+  val sim = Module(new SimUART(dataWidth, uartno))
+
+  sim.io.clock := clock
+  sim.io.reset := reset
+
+  sim.io.serial.out.bits := txfifo.io.deq.bits
+  sim.io.serial.out.valid := txfifo.io.deq.valid
+  txfifo.io.deq.ready := sim.io.serial.out.ready
+
+  rxfifo.io.enq.bits := sim.io.serial.in.bits
+  rxfifo.io.enq.valid := sim.io.serial.in.valid
+  sim.io.serial.in.ready := rxfifo.io.enq.ready
+}
+
+class SimUART(width: Int, uartno: Int) extends BlackBox(Map("WIDTH" -> IntParam(width), "UARTNO" -> IntParam(uartno))) with HasBlackBoxResource {
+  val io = IO(new Bundle {
+    val clock = Input(Clock())
+    val reset = Input(Bool())
+
+    val serial = Flipped(new SerialIO(width))
+  })
+
+  addResource("/testchipip/vsrc/SimUART.v")
+  addResource("/testchipip/csrc/SimUART.cc")
+  addResource("/testchipip/csrc/uart.cc")
+  addResource("/testchipip/csrc/uart.h")
+}
+
+trait HasPeripheryUARTWithAdapter extends HasPeripheryUART { this: BaseSubsystem =>
```

Comment:
This should be named `CanHavePeripheryUartAdapter`

Reply:
Ok changed them to `CanHave`s. Note, the name of the trait is "have UART with Adapter". not "have UART adapter" to note that this adds both the UART and the corresponding adapter. The empty trait is to add the main `HasPeripheryUART` trait while having similar naming to the `Imp` trait.